sdk/ai/azure-ai-projects/azure/ai/projects/models/_patch.py

```diff
@@ -960,16 +974,202 @@ def on_unhandled_event(self, event_type: str, event_data: Any) -> None:
+            
+        elif event_type == AgentStreamEvent.THREAD_RUN_STEP_DELTA.value:
+            iterator_items.append((event_type, _safe_instantiate(RunStepDeltaChunk, parsed_data)))
+        elif event_type == AgentStreamEvent.DONE.value:
+            iterator_items.append((event_type, None))
+        else:
+            iterator_items.append((event_type, ThreadError(error=str(parsed_data))))
+        return iterator_items
+        
+        
+    
+class BaseAsyncAgentEventHandler(AsyncIterator[Tuple[str, T]], Generic[T]):
+    def __init__(
+        self
+    ):
+        self.done = False
+        self.buffer = ""
+
+    def _init(
+        self,
+        response_iterator: AsyncIterator[bytes],
+        submit_tool_outputs: Callable[[ThreadRun, "BaseAsyncAgentEventHandler"], Awaitable[None]],
+        event_handler: Optional["BaseAsyncAgentEventHandler"] = None,
+    ):
+        self.response_iterator = response_iterator
+        self.event_handler = event_handler
+        self.submit_tool_outputs = submit_tool_outputs
+            
+    async def __anext__(self) -> Tuple[str, T]:
+        while True:
+            try:
+                chunk = await self.response_iterator.__anext__()
+                self.buffer += chunk.decode("utf-8")
+            except StopAsyncIteration:
+                if self.buffer:
+                    event_data_str, self.buffer = self.buffer, ""
+                    if event_data_str:
+                        return await self._process_event(event_data_str)
+                raise
+
+            while "\n\n" in self.buffer:
+                event_data_str, self.buffer = self.buffer.split("\n\n", 1)
+                return await self._process_event(event_data_str)
+            
+    
+    async def _process_event(self, event_data_str: str) -> Tuple[str, T]:
+        raise NotImplementedError("This method needs to be implemented.")
+    
+    async def until_done(self) -> None:
+        """
+        Iterates through all events until the stream is marked as done.
+        Calls the provided callback function with each event data.
+        """
+        try:
+            async for _, _ in self:
+                pass
+        except StopAsyncIteration:
+            pass
 
+    
+    async def yield_until_done(self, yield_callback: Callable[[str, T], Awaitable[Optional[R]]], **kwargs) -> AsyncGenerator[R, None]:
+        """
+        Iterates through all events until the stream is marked as done.
+        Calls the provided callback function with each event data.
+        """
+        try:
+            async for event_type, event_obj in self:
+                to_be_yield = await yield_callback(event_type, event_obj, **kwargs)
+                if to_be_yield:
+                    yield to_be_yield
+        except StopAsyncIteration:
+            pass
 
-class AsyncAgentEventHandler:
+    
+class DefaultAsyncAgentEventHandler(BaseAsyncAgentEventHandler[StreamEventData]):
```

Comment:
I would like to keep AsyncAgentEventHandler name because that was originally also default

Reply:
Done